cpp/src/slice2cs/Gen.cpp

```diff
@@ -1776,12 +1798,32 @@ Slice::Gen::TypesVisitor::visitStructStart(const StructPtr& p)
     emitDeprecate(p, false, _out);
     emitCommonAttributes();
     emitCustomAttributes(p);
+
+    bool equatable = isEquatable(p);
+
+    if (!equatable)
+    {
+        _out << nl << "[global::System.Diagnostics.CodeAnalysis.SuppressMessage(";
+        _out.inc();
+        _out << nl << "\"Microsoft.Performance\","
+            << nl << "\"CA1815: Override equals and operator equals on value types\","
+            << nl << "Justification=\"struct with sequence or dictionary field needs custom Equals and GetHashCode\")]";
+        _out.dec();
+    }
```

Comment:
Why do we suppress this warning, wouldn't it be better if the user sees it and provides an implementation in a partial class?

Reply:
You are right, the suppression of this warning goes against the rule.

I started down the path of providing IEquatable implementations to all non-IEquatable generated structs. It's doable but:
 - it's a pain to write all this code
 - it's error-prone: you or I can easily mess-up this hand-generated implementation, and if someone modifies the Slice definition, there is a good chance he or she won't update the partial struct IEquatable implementation
 - in practice, it means we need to make DictionaryExtensions public to implement Glacier2.SSLInfo's IEquatable and various IceGrid structs
 - sometimes, these never-used IEquatable implementations will be the only hand-written code for an assembly, e.g. IceGrid

Since all these IEquatable implementations are never used, they are never tested, which is not great. And obviously their performance does not matter since they are never used.

An alternative to a full implementation (since they are never used) would be to suppress the CA1815 warning with a partial struct only for this purpose (I haven't tried, but I assume it works). However that's not great either: you have simple public struct with a sequence<string>, you call Equals on it and get object equality because the implementer didn't expect anybody to ever use Equals on it ... not nice.

So I think we should make all generated structs implement IEquatable, using deep equality for sequence and dictionary members. And possibly provide a special metadata like `[cs:custom-equality]` that skips the generation of Equals and GetHashCode (but not IEquatable, == and other boilerplate) so that the user can provide a custom implementation with a different semantics / hash code caching etc.

Reply:
> An alternative to a full implementation (since they are never used) would be to suppress the CA1815 warning with a partial struct only for this purpose (I haven't tried, but I assume it works). However that's not great either: you have simple public struct with a sequence, you call Equals on it and get object equality because the implementer didn't expect anybody to ever use Equals on it ... not nice.

We support defining cs:attribute in Slice, no need to add a partial class.

> So I think we should make all generated structs implement IEquatable, using deep equality for sequence and dictionary members. And possibly provide a special metadata like [cs:custom-equality] that skips the generation of Equals and GetHashCode (but not IEquatable, == and other boilerplate) so that the user can provide a custom implementation with a different semantics / hash code caching etc.

Sounds fine